### DISCUSSION TITLE
Getting Mantis Gamepad Pro to work in Waydroid

### DISCUSSION THREAD

**Discussion by appleinario** [waydroid/waydroid#605] [Q&A]
I have seen on a thread in Reddit where a person was able to get Mantis working on Waydroid via root (see: https://www.reddit.com/r/linux_gaming/comments/tkgsnc/honkai_impact_3rd_running_on_waydroid_with/)

I would like to know how I can achieve the same results? There are three ways of getting Mantis Gamepad Pro to work, and I have failed on all three of them:

1. Launching MantisBuddy Directly On-the-Device (Android 11 & Above)
- This one seems to be the ideal solution, however, I am stuck on the step where it asks the user to enable Wireless Debugging. In order to enable Wireless Debugging, the system asks the user to enable Wifi Connections, which doesn't seem to work with Waydroid. I have tried tricking Wireless Debugging via `waydroid prop set persist.waydroid.fake_wifi com.android.settings` to no avail. I have yet to see a workaround for this.

2. Android-to-Android
- My Android, connected to my computer, doesn't communicate with Waydroid.

3. Android-to-PC
- The code which the website provides (see: https://mantispro.app/buddy/) are only available in Windows and Mac.

**Comment by selurvedu** [comment]
Works for me in Android-to-Android mode on rooted Waydroid (Magisk Delta Canary). Here's how I did it:

1. I selected Android-to-Android mode.
2. A superuser prompt popped up.
3. I granted the root access forever.
4. Mantis automatically set itself up right away. I didn't have to connect another device or issue any adb commands.
5. I enabled the direct controller access by using the commands that YogSottot provided in this [comment](https://github.com/waydroid/waydroid/issues/289#issuecomment-1530761751):
```shell
waydroid prop set persist.waydroid.udev true
waydroid prop set persist.waydroid.uevent true
```
This has to be done only once.
6. I reconnected the gamepad.
7. From this point on it Just Works™.

You may also try to use the command from the comment I mentioned in case this doesn't work for you or if you don't want to enable root access, however, you would have to issue that command every time you restart Mantis. This is very similar to how you start XtMapper (I would be happy to switch to it, but [currently it does not work on Waydroid](https://github.com/Xtr126/XtMapper/issues/35)).

Hope this helps.

----

P.S. If you're not rooted and decide to install Magisk, you can use [waydroid_script](https://github.com/casualsnek/waydroid_script) for easy installation. That's what I used. A manual installation guide and some additional hints are available here: <https://github.com/pagkly/MagiskOnWaydroid>. There's one important thing though, the manual states that Zygisk isn't working yet. It does work already, but as August of 2023, it is only available on the Canary release channel (version `fdac22ba-delta (25210)`), so you may want to install it / switch to it right away if you intend to install any Magisk modules. This is not required if you only need it for Mantis, since Mantis only needs `su`, and that works on any release channel.

**Reply by gagantous** [reply]
hi, can you elaborate more in step 2 . A superuser prompt popped up ? there is no prompt popped up when i try it

**Reply by selurvedu** [reply]
@gagantous sure!

They changed it in a recent update. Here's how it works for me:

1. In Mantis, press the game icon.
<details><summary>Screenshot</summary>

![image](https://github.com/waydroid/waydroid/assets/11262062/3071850f-0166-428d-825b-78095169a496)
</details>

2. Press the "Start" button next to "Mantis Buddy".
<details><summary>Screenshot</summary>

![image](https://github.com/waydroid/waydroid/assets/11262062/a1bff64b-4eba-4d75-ba0f-8d116adf003d)
</details>

3. Press "Connect Mantis Buddy via another Android"
<details><summary>Screenshot</summary>

![image](https://github.com/waydroid/waydroid/assets/11262062/55b9e685-1d0b-4886-9ad6-5bd58d50aefa)
</details>

4. When the Superuser Request pops up, make sure "Forever" is selected and press "Grant".

<details><summary>Screenshot</summary>

![image](https://github.com/waydroid/waydroid/assets/11262062/324d545c-76df-4b53-a0a2-74b4dc33c0ea)
</details>

5. You should see "Mantis Buddy Connected. You're all set."
6. Press Back twice and press "Launch game". That's it.

Alternatively, press the :gear: button (the gear / cog icon), then press the "Start" button next to "Mantis Buddy", then continue from step 3.

<details><summary>Screenshot</summary>

![image](https://github.com/waydroid/waydroid/assets/11262062/6de3f7ec-05fe-4019-9611-cba42063912d)
</details>

If you still don't get the superuser prompt, make sure Magisk is installed correctly and the root permissions work for other apps.

Hope this helps!
